_posts/2024-02-19-jco-1.0.md

```diff
@@ -0,0 +1,270 @@
+        s.into()
+    }
+}
+```
+
+Now if we re-run `cargo component check`, it should compile without any
+warnings. That means we're ready to run:
+
+```bash
+$ cargo component build      # Compile our Rust library to a Wasm Component
+```
+
+This will have written the generated Wasm code to
+`target/wasm32-wasi/debug/playground_jco.wasm`. Once Rust lands support for a
+native Preview 2 target, expect this to change to
+`target/wasm32-wasi-p2/debug/playground_jco.wasm` or similar, as the current set
+of workarounds will no longer be needed (important if you're reading this
+anytime after March 2024 or so).
+
+### Writing a JavaScript binary
+
+Okay, it's time to use `jco`. First we want to take our Rust code, and generate
+JS bindings from it. We can do this using the `jco transpile command`. Let's
+point it at our newly generated Wasm library, like so:
+
+```bash
+$ jco transpile \                                   # Call jco transpile
+    target/wasm32-wasi/debug/playground_jco.wasm \  # Point it at our Rust library
+    -o target/jco                                   # Write the result to `target/jco`
+```
+
+On success it will tell you it's generated a number of JS files as part of
+its CLI output:
+
+```text
+Transpiled JS Component Files:
+
+ - target/jco/interfaces/wasi-cli-environment.d.ts      0.09 KiB
+ - target/jco/interfaces/wasi-cli-exit.d.ts             0.16 KiB
+ - target/jco/interfaces/wasi-cli-stderr.d.ts           0.17 KiB
+ - target/jco/interfaces/wasi-cli-stdin.d.ts            0.17 KiB
+ - target/jco/interfaces/wasi-cli-stdout.d.ts           0.17 KiB
+ - target/jco/interfaces/wasi-clocks-wall-clock.d.ts    0.11 KiB
+ - target/jco/interfaces/wasi-filesystem-preopens.d.ts   0.2 KiB
+ - target/jco/interfaces/wasi-filesystem-types.d.ts     2.71 KiB
+ - target/jco/interfaces/wasi-io-error.d.ts             0.08 KiB
+ - target/jco/interfaces/wasi-io-streams.d.ts           0.58 KiB
+ - target/jco/playground_jco.core.wasm                  1.83 MiB
+ - target/jco/playground_jco.core2.wasm                 16.5 KiB
+ - target/jco/playground_jco.d.ts                       0.72 KiB
+ - target/jco/playground_jco.js                           40 KiB
+```
+
+The next step is to import this into our Node.js file. In order to do this, we
+first have to update our `package.json` to declare that we're writing a JS
+module. In our package.json, add the following line:
+
+```json
+{
+  "type": "module",
+}
+```
+
+And then finally, we're ready to write some JS code. we're going to import the
+`scream` function from our Rust library, pass it a string, and then print the
+output of the string.
+
+```js
+import { scream } from "./target/jco/playground_jco.js";
+
+let msg = scream("chashu would like some tuna");
+console.log(msg);
+```
+
+Now if we run this using `node index.js`, we should get the following output:
+
+```text
+CHASHU WOULD LIKE SOME TUNA!!1!
+```
+
+And with that, we've successfully written our first Rust component for Node.js.
```

Comment:
Very important, to me at least, is the idea that you can swap out ANY language component here that implements the same interface. So you should mention the variations that the component model gives you here: that you could implement the same WIT definition in Go and the same index.js file would work; that the same workflow would work for JavaScript, and so on. 

Up to you.

Reply:
good call; done!